include/jakalib/math.h

```diff
@@ -0,0 +1,42 @@
+#ifndef JAKALIB_MATH_H
+#define JAKALIB_MATH_H
+
+#include <algorithm>
+#include <cmath>
+
+namespace jakalib {
+namespace {
+template <typename T>
+T sqr(T x) {
+  return x * x;
+}
+
+template <typename T>
+T hypot3(T x, T y, T z) {
+  x = std::abs(x);
+  y = std::abs(y);
+  z = std::abs(z);
+
+  T m = std::max(std::max(x, y), z);
+  if (m == T(0)) {
+    return T(0);
+  }
+  return m * std::sqrt(sqr(x / m) + sqr(y / m) + sqr(z / m));
```

Comment:
division is more expensive than multiplication and (often? always?) cannot be optimized automatically. Consider making a second temporary to hold `1/m`.